### PR TITLE
feat: image orchestrator model ladder (v0.3.0)

### PR DESCRIPTION
**Pull request by KingGyuSuh**
## Summary

Image generation is a codex agent turn (the "orchestrator") that calls the built-in image tool. A stronger orchestrator model/effort interprets prompts, conditions references, and holds the requested aspect more reliably — but hardcoding a model would break accounts that can't access it (e.g. ChatGPT Free). This PR adds a catalog-driven preference ladder, released as **v0.3.0**.

**Stacked on #8** (the v0.2.1 `--sandbox workspace-write` fix). This PR was restacked so it contains only the ladder; the sandbox migration, `approval_policy` pin, `status` fixes, and version alignment now live in #8. Merge #8 first — GitHub retargets this PR to `main` automatically. The previous four-commit history is kept locally under the tag `backup/pr4-before-restack`.

## What changed

- The wrapper probes the account's live catalog via `codex debug models` (keeps `visibility: "list"` models and their `supported_reasoning_levels`), bounded by a 15s timeout, and picks the first available rung of **`gpt-5.6-luna` high → `gpt-5.6-terra` medium → `gpt-5.6-sol` high → `gpt-5.6-sol` low**, passing it as `-m <model> -c model_reasoning_effort="<effort>"`.
- A failed probe or no matching rung adds **no flags** — codex falls back to its config default, so generation never regresses.
- If the backend rejects a ladder-selected model at turn start (e.g. a custom `model_provider` that doesn't serve official slugs the catalog still lists), the wrapper retries once without `-m` / `-c`. The rejection happens before any image tokens are spent; unrelated failures (quota, auth, transport) are not retried.
- `CODEX_IMAGE_MODEL` + `CODEX_IMAGE_EFFORT` (both required, else a hard error) force a specific pair, skip the probe, and never fall back.
- `/codex-image:status` reports the resolved orchestrator and only fails readiness on an inconsistent override.
- Docs: `docs/ARCHITECTURE.md` "Image orchestrator model ladder" load-bearing section, README "Image model" section, CONTRIBUTING, CHANGELOG `[0.3.0]`.

## Type of change

- [ ] Bug fix
- [x] New feature
- [ ] Refactor / cleanup
- [x] Documentation
- [ ] CI / build

## Checklist

- [x] `npm test` passes — 40/40 (22 ladder tests on top of #8's 18)
- [x] `npm run validate:plugin` passes
- [x] `docs/ARCHITECTURE.md` "Load-bearing edge cases" updated in the same PR
- [x] `CHANGELOG.md` updated (`[0.3.0]`)
- [x] Scope respects "image generation only"

## Codex review (gpt-6-astra, `codex exec review --commit`)

Four headless review rounds on this branch, each on Codex CLI 0.154.0:

1. Ladder commit — P2: the fallback classifier ran over the whole stderr tail, and `codex exec` echoes the prompt, agent messages, and tool output to stderr, so a prompt mentioning "Unknown Model" plus an unrelated failure could trigger a second (billed) image turn. Confirmed live; fixed by classifying only codex's own `ERROR:` record and refusing to retry once the turn has started.
2. Follow-up 1 — three P2s where that fix suppressed a *valid* fallback: multi-line (pretty-printed) `ERROR:` records, ANSI-styled `ERROR:` under `FORCE_COLOR`, and a prompt containing a standalone `codex` / `exec` line. Fixed by cutting the echoed instruction off first, stripping ANSI, and reading the record from the first `ERROR:` line to the end.
3. Follow-up 2 — P2: with the bounded tail, a last-line echo heuristic could match a later shell command ending in the same text and erase turn-start evidence. Fixed by a streaming stderr monitor with a sticky `started` flag over the whole stream (`createStderrMonitor`).
4. Follow-up 3 — no actionable regressions.

Live on 0.154.0 after each round: a bogus ladder model with a prompt containing "Unknown Model", bare `codex` / `exec` lines and an output path is rejected at startup and retried once with the config default (exit 0); a normal ladder generate still saves a 512×512 PNG.

## Test plan

On Codex CLI **0.154.0** (restacked branch, on top of #8):

- `/codex-image:status` → `Ready: yes`, `OK Image orchestrator: gpt-5.6-luna (effort high) [ladder]`.
- Live `generate` in a non-git scratch directory → stderr `codex image orchestrator: gpt-5.6-luna (effort high) [ladder]`, session header `model: gpt-5.6-luna`, `reasoning effort: high`, `approval: never`, `sandbox: workspace-write [workdir, /tmp, $TMPDIR]`, a `SAVED:` line, 512×512 PNG on disk, exit 0.

Earlier runs on Codex CLI 0.144.5 (before the restack; same ladder code): plain generate, `--ref` generate, edit (input untouched, md5-verified), env-override generate (`gpt-5.6-terra` / `medium`), and a bogus-model rejection captured verbatim for `isOrchestratorRejection`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01QY31z6pZhrYW7AcZknjMNP

### CODE REVIEW

File: scripts/codex-image.mjs [hunk 1]
```diff
@@ -388,10 +584,25 @@ async function handleGenerate(argv) {
...
+  if (orchestrator) {
+    console.error(
+      `codex image orchestrator: ${orchestrator.model} (effort ${orchestrator.effort}) [${orchestrator.source}]`
+    );
+  }
   const codexArgs = [
     "exec",
-    "--full-auto",
+    "--sandbox",
+    "workspace-write",
```
[the remark this answers is not in the document]
Comment (KingGyuSuh) [reply scripts/codex-image.mjs]: Verified and fixed in deec3fc. Reproduced on 0.144.5 with `--ignore-user-config` isolation: `--full-auto` + `approvals_reviewer="auto_review"` keeps `approval: never`, bare `--sandbox workspace-write` + auto_review flips the header to `approval: on-request` (exactly the `build_exec_config` rebuild path cited), and adding `-c approval_policy="never"` restores `approval: never` while being a no-op for non-auto_review configs. The wrapper now passes the explicit override via a shared `CODEX_EXEC_BASE_ARGS`, with a unit test pinning the contract and the load-bearing docs updated.

File: scripts/codex-image.mjs [hunk 2]
```diff
@@ -130,6 +164,156 @@ function findImagegenSkill() {
...
+}
+
+// Query the account's live model catalog. Best-effort and side-effect free: any
+// failure (codex missing, not logged in, offline) returns null and the caller uses
+// the codex config default rather than failing the generate/edit.
+function loadModelCatalog(cwd) {
+  const result = runSync(CODEX.command, [...CODEX.prefix, "debug", "models"], {
+    cwd,
+    maxBuffer: 32 * 1024 * 1024
+  });
```
[the remark this answers is not in the document]
Comment (KingGyuSuh) [reply scripts/codex-image.mjs:265]: Fixed in 75f239e: the probe now runs with a 15s `spawnSync` timeout, and a timed-out/killed probe reports a null status which the existing `status !== 0` check already treats as an unavailable catalog — generation proceeds on the Codex config default.

File: scripts/codex-image.mjs [hunk 3]
```diff
@@ -130,6 +181,156 @@ function findImagegenSkill() {
...
+// Walk the ladder top-to-bottom; return the first rung whose model is in the catalog
+// and whose effort that model supports. null if the catalog is missing or no rung fits.
+function selectOrchestratorFromLadder(catalog, ladder = CODEX_IMAGE_ORCHESTRATOR_LADDER) {
+  if (!catalog) {
+    return null;
+  }
+  for (const rung of ladder) {
+    const efforts = catalog.get(rung.model);
+    if (efforts && efforts.has(rung.effort)) {
+      return { model: rung.model, effort: rung.effort };
```
[the remark this answers is not in the document]
Comment (KingGyuSuh) [reply scripts/codex-image.mjs:281]: Addressed in 75f239e with a different mechanism than suggested: instead of detecting custom providers up front (fragile, and permanently loses the ladder benefit for those users), the ladder pick is now best-effort at spawn time. A backend rejection of the `-m` slug kills the turn at startup — verified live on 0.144.5: HTTP 400 `"The 'X' model is not supported…"` on stderr with empty stdout, i.e. before any image tokens are spent — so the wrapper matches that signature (`isOrchestratorRejection`, only on non-zero exit) and retries once without `-m`/`-c`, landing on the configured default that worked pre-ladder. Unrelated failures (quota/auth/transport) don't match and are never retried, so no double-billing. Env-forced pairs (`CODEX_IMAGE_MODEL`/`CODEX_IMAGE_EFFORT`) deliberately never fall back — an explicit assertion should surface its error. Live-verified: bogus ladder model → rejection → retry → config default → success.
